Create basic template rendering engine, test and example

hems opened the pull request:
This PR is a working prototype that will:

 - Take a folder with multiple [handlebars](https://handlebarsjs.com/) templates
 - Render all of them using provided data
 - Output rendered files to an output folder

So far:
 - A basic [README.md](https://github.com/hems/bfx-api-client-generator/blob/master/README.md) with travis integration was created
 - A [simple mocha test](https://github.com/hems/bfx-api-client-generator/blob/master/tests/index.js) was created
 - I started creating [examples](https://github.com/hems/bfx-api-client-generator/tree/master/examples) in order to understand better what features are required in order to generate all the needed clients

If we need to create files dynamically, as opposed to only populate template files, would be handy to __hooks__ so _data_ and _files_ can be processed in the different stages of the pipeline.

At the moment i'm forking one process for each processed file, offloading each template to [a simple worker](https://github.com/hems/bfx-api-client-generator/blob/master/src/worker.js), hopefully improving the performance when dealing with more complex templates.

Source was linted using [standard](https://standardjs.com/) as per @f3rno recommendation

hems committed:
first version with basic test

hems committed:
removing console log, updates package.json and README.md

hems committed:
what about we use all the cpus?

hems committed:
improve file_list function

f3rno commented on src/lib/fork_job.js:
```diff
@@ -0,0 +1,24 @@
+    if (code === 0) return deferred.resolve()
+
+    // something went wrong!
+    if (code !== 0) return deferred.reject(code)
```
Just return a reject here if the code is non-zero, and return the resolve otherwise (1 less `if`), but it's a minor detail

hems replied:
yeah return home early!

f3rno commented on src/lib/promises/create.js:
```diff
@@ -0,0 +1,12 @@
+    reject = arguments[1]
+  })
+
+  return { promise, resolve, reject }
```
This is really cool :heart:

hems replied:
💃

f3rno commented on src/lib/read.js:
```diff
@@ -0,0 +1,10 @@
+ * Reads given path and return it's content as utf-8
+ */
+const P = require('bluebird')
+// shall we use something like require fs/promises ?
```
Vestigial comment? Or what do you mean?

hems replied:
it seems you don't need to promisify the "fs" package as on newer node there is "fs/promises" but i'm not sure, i better delete this indeed

f3rno commented on tests/index.js:
```diff
@@ -0,0 +1,48 @@
+const read = require('../src/lib/read')
+
+const EXAMPLES_PATH = path.join(__dirname, '/../examples/')
+const OUTPUT_PATH = path.join(__dirname, '/../output/')
```
Let's not pollute the root folder with rendered test templates, but otherwise this looks fine for now.

hems replied:
yeah that's right, i was already thinking about moving this to another place. will do it on the next version

f3rno commented on package.json:
```diff
@@ -0,0 +1,30 @@
+  "version": "0.0.1",
+  "main": "src/generator.js",
+  "author": "hems <hems.inlet@gmail.com>",
+  "contributors":[],
```
Missing space before `[]` (`npm i` corrects it & it causes a new diff)

hems replied:
+1

f3rno commented on src/generator.js:
```diff
@@ -0,0 +1,37 @@
+  const files = await ls(templatePath)
+
+  // create one worker for each file
+  await P.map(files, async (file) => {
```
This is a great starting point, but keep in mind we will need a way to compose/nest templates/block helpers. i.e, for the REST API, we can write a single block helper implementing  `_makeAuthRequest()`, and then use it in a separate `rest2.js.hbs` template file. This should scale to nesting arbitrary templates inside of each other (no need to go deeper than 2 levels imo, so root + 1 more), not just block helpers. This may require a pre-rendering stage.

hems replied:
yeah i kept that in mind and i think i got a very elegant idea on how to solve that, will let you know when i get it down!

f3rno commented on src/generator.js:
```diff
@@ -0,0 +1,37 @@
+    await fork(worker, job)
+
+  // use as many cpus as possible
+  }, {concurrency: os.cpus().length})
```
Very nice :muscle:

hems replied:
👽

f3rno reviewed (changes requested):
Overall this is a great starting point; very nice usage of async :slightly_smiling_face: 

For the initial merge, I'd like to see the architecture for a pair of REST2 block helpers (public/auth) & real-world usage with endpoints. We can leave the method doc headers for the next iteration, once we have a matching implementation in another language (I'll push python or ruby first).

Also, keep in mind we will want to maintain 1 master template for all of the different RESTv2 implementations; this means one master list of URLs/arguments (query & body). One way of doing this might be to dynamically select block helpers when rendering the templates, and structuring them so they are the same across all languages. If this is not possible we can split it up, but rest2 should be doable (ws2 may get complicated).

hems commented:
Regarding "Also, keep in mind we will want to maintain 1 master template for all of the different RESTv2 implementations; this means one master list of URLs/arguments (query & body)", cool, i'll keep it in mind.

Will add the libraries as examples ( and also write some tests for it ) on the repository so we can play with real world problems

hems committed:
moving output folder to inside example folder, simple syntax improvement

hems committed:
adding data as function and data as promise examples,
adds tests and refactor examples

hems committed:
adds inline and external partial examples

hems committed:
started templating rest2.js

hems commented:
@f3rno i believe i have addressed your concern "but keep in mind we will need a way to compose/nest templates/block helpers" by adding the partials examples on the repository.

In order to start dealing with the actual problem we need to solve and not with the generator architecture i have also started templating rest2.js so I can start modeling the actual specification file according to our needs.

In the meanwhile i'm using inline templates on rest2.js, but in the future we could extract those into external files and use them as external partials ( as i have done in one of the examples ).

One cool feature I have added is the possibility to use a function or promise as specification, so if we need to generate specification dynamically, fetch an API to generate the spec or anything async it would already be possible.

Tests were updated accordingly.

hems committed:
remove unused example from the partials section
fix partials examples link
improve documentation

hems committed:
compiling node, ruby and go at the same time
 - scaffolded test
 - improved performance ( removed multi threading )
 - simplified tests and examples

hems commented:
@f3rno So far:

 - I believe so far I have addressed your concern "but keep in mind we will need a way to compose/nest templates/block helpers" by adding the partials examples. I can look into blocks if/when we need

 - Added ruby, node and go library [side by side](https://github.com/hems/bfx-api-client-generator/tree/master/examples/bitfinex-clients)

 - Compile time is much faster without multi-threading ( fork_child ), 🤣 ( reminds me: overoptimising is always wrong )

 - Since most files on the repositories are templates and compiling time is super fast, we might as well not use the file extension __.hbs__ and just assume they are all templates

 - I [started templating rest2.js](https://github.com/hems/bfx-api-client-generator/blob/master/examples/bitfinex-clients/templates/node/lib/transports/rest2.js) on the node.js library, using inline partials for now, but surely will organise this better in the future

 - I noticed the spec will have to be improved a lot to cover all 3 libraries, suggestions are welcome 

 - Ideally, as you mentioned we would be able to create files dynamically as a more convient way of generating the libraries compared to using blocks / partials. I'll have a look into implementing hooks so we can temper the pipeline to generate those extra files during compile time or as you said pre-compile templates.

hems committed:
Implements node.js model generation based on spec

hems committed:
template for node/lib/model/index.js

hems committed:
adds json helper

hems commented:
@f3rno I went a bit further with the node.js templating:

 - done a [simple spec for the node.js models](https://github.com/hems/bfx-api-client-generator/blob/master/examples/bitfinex-clients/node.models.js)

 - templated [lib/models/index.js](https://github.com/hems/bfx-api-client-generator/blob/master/examples/bitfinex-clients/templates/node/lib/models/index.js)

 - created a [couple of partials](https://github.com/hems/bfx-api-client-generator/tree/master/examples/bitfinex-clients/partials) needed for most of the models

 - Noticed some models have some exceptions, some of them have extra fields, some of them are a totally different thing, such as __Order__ and __OrderBook__

I'm still unsure of how the final spec will be but I'm planning to have only 1 specification that will fit all client libraries.

Potentially there will be some specific pre-processing of the specification for each language.

For the node.js Models we could even create the files for most of them dynamically if we wanted to, the only problem would be the more complicated models such as Order and OrderBook

hems committed:
templating node.js test files

hems committed:
replacing api urls
